### DISCUSSION TITLE
[Feature request] Discussing the auto self updater in Cmder

### DISCUSSION THREAD

**Discussion by mattdkerr** [cmderdev/cmder#3066] [General]
# Branched discussion from https://github.com/cmderdev/cmder/issues/2927#issuecomment-3902765581

- for the original fzf integration request, see: https://github.com/cmderdev/cmder/issues/3067
- for auto self updater discussion, see below

---

Also requested in https://github.com/orgs/cmderdev/discussions/2875

**Comment by chrisant996** [comment]
> [...] I know I would appreciate a way that would allow the users to update their `clink-completions` package the same way that they can just do `clink update` and enjoy the latest version of `clink`. (Great update, by the way, @chrisant996).

It's possible to make an auto-updater for clink-completions inside Cmder.  If a script updater is on by default, then a side effect is that it will automatically silently discard any local edits that users have made to the scripts.

It's also possible to make a script that can update clink-completions.  It's just a matter of extracting the contents from the latest .zip release file into the vendor/clink-completions directory.

**Comment by DRSDavidSoft** [comment]
> If a script updater is on by default, then a side effect is that it will automatically silently discard any local edits that users have made to the scripts.

That's a non-issue in my opinion since we have instructed users for many versions now (since a couple of years, I believe) _NOT TO_ modify scripts in the `vendor` directory as they'll be overwritten in newer versions. @daxgames even took great care to move any files with potential configuration to the `config` directory to avoid losing modifications on each update. In any case, I believe the standard and recommended way that is suggested to update Cmder is to simply extract the ZIP file directly over the previous version which essentially has the same effect, which means it would be a good idea if the auto updater script would do this as well.

This is off topic, but may I ask that you'd add this functionality directly in either `clink` or `clink-completions`, since it's already a working feature of `clink`? That way, we could include and use it in Cmder, as well.

**Comment by chrisant996** [comment]
> Do we need to put the scripts in a directory that won't be loaded by default, and either a) ask the user to move the files into another directory, or b) change the configuration to load scripts from the specified path? I'm under the assumption that there are no config files that are used to disable/enable scripts that are loaded by clink (sorry if that's the case).

Cmder needs to [do what Cmder does for clink-completions](https://github.com/cmderdev/cmder/blob/4daff054ebe5ea2d28b8e60a19681a8cd1f1e72d/vendor/clink.lua#L682-L693) (please follow the link to see what Cmder already does).

The clink-completions directory is a custom directory, and Cmder has Lua code that specifically loads each of the scripts in that directory.  Cmder can continue to use the same approach for any other Lua script collections that Cmder wants to include.

> may I ask that you'd add this functionality directly in either `clink` or `clink-completions`, since it's already a working feature of `clink`? That way, we could include and use it in Cmder, as well.

I welcome the question, but the answer has to be "no" about building it into Clink, because:

1. Neither clink-completions nor clink-gizmos are part of Clink.  They aren't features of Clink, they're separate things which depend on Clink and add enhancements to Clink.  (Like how a user's personal .bat scripts aren't a feature of CMD, they're a feature of the user's personal configuration.)
2. Clink's updater is for Clink.  It shouldn't update things other than Clink; that would be a recipe for trouble.
3. The clink-completions and clink-gizmos repos already have an update mechanism.  Their readme files recommend to install using git, and `git pull` is the update command (or other standard git workflows such as upstreams and branches).
4. Cmder has a custom installer for the scripts, in effect.  An updater for Cmder's scripts would need to be part of Cmder.

**Comment by chrisant996** [comment]
How much interest is there in having an auto-updater for clink-completions and/or clink-gizmos?  I'm currently leaning against investing the work for that, but I'm very open to discussion and debate.

<br/>

> [!CAUTION]
> **An auto-updater for clink-completions or clink-gizmos sounds great, but has some important issues:**

<br/>

First, any local changes the user has made to the scripts will get overwritten.  That isn't necessarily a reason against having an auto-updater, but it's something that needs to be communicated clearly to users.

Second, it's critical that auto-updaters need to validate digital signatures before installing updates.  This is not just a theoretical attack vector; other software has been attacked in exactly this way.  I changed Clink's auto-updater in v1.9.12 to validate digital signatures before installing an update.
  - Some software security attacks in recent months have exploited auto-updaters (you may have heard of the Notepad++ woes, among other recent cases).
  - If the network connection can be compromised (e.g. man-in-the-middle network attack), or if the file server is hacked (e.g. if GitHub gets hacked), or if the author's account is hacked (e.g. if Validimir Kotikov's account or my account are hacked), then it becomes possible for the auto-updater to receive a malicious/fake update package.
  - Digital signature validation is the solution, provided that the private key is properly secured.  Notepad++ put its private key in its public repo, making it possible (and simple!) for anyone to sign updates maliciously.  In contrast, my private key exists exclusively only on a special physical security device (an ACR39T rev. B from Advanced Card Systems, Ltd.) which has to be physically plugged in while signing, and requires a PIN to access it.

<br/>

So, if an auto-updater is desired, then two things would need to happen:
1. A custom updater program (or script) would need to be created and validate digital signatures.
    - Clink v1.9.12 and newer have an undocumented Lua API that can validate digital signatures (Clink's own updater uses it).  So the validation itself is actually the easy part.
    - It's a medium-sized chunk of work to create a robust auto-updater, and I'm not sure whether I want to sign up for doing that.
2. The publisher of clink-completions and clink-gizmos (i.e. me) would need to generate a digitally signed Catalog file for each update.
    - It's a small-sized chunk of work to create a script to automate creating a signed Catalog file for each update.  I'm willing to add that to my workflow, but it doesn't accomplish anything without a custom updater program to make use of the signed Catalog file.

**Comment by DRSDavidSoft** [comment]
> [!NOTE]
> I split the discussion about auto self update mechanism, from the previous fzf integration tracking issue to better keep things organized and stay on topic. 
> This new page is dedicated to discussing this feature.

---

> How much interest is there in having an auto-updater for clink-completions and/or clink-gizmos?

> I'm currently leaning against investing the work for that, but I'm very open to discussion and debate.

@chrisant996 

I'm personally quite invested in implementing a full self-updater mechanism baked directly into Cmder, as I've previously proposed as part of https://github.com/cmderdev/cmder/issues/2788#issuecomment-2543380699.

My reasoning is that Cmder developers often update the core mechanisms in Cmder itself (including the [vendor/lib/lib_*.cmd](https://github.com/cmderdev/cmder/tree/master/vendor/lib) files, PS1 scripts, and more) -- and currently there is no system in place for the users to benefit from these updates the same way they could benefit from updates to clink, using a simple `clink update` command.  Same with any updates made to the clink-completions package. In order to benefit from updates to either, they would need to manually download a ZIP package and extract it in the correct place, or worse, download a new Cmder package each time they want to update (which also has the downside of not being regularly released with the latest updates comitted to the repo)

As a user of Cmder myself, every time that I see a new clink update is released and perform "clink update", it is sad to see that all the rest of the Cmder package gets left behind.  It also feels inconsistent.

For this reason, I've been considering adding a "cmder update" functionality, script, or maybe something external that can be invoked by the user to update any outdated Cmder installation, quickly and correctly to the latest version available in the repo. It should be safe, fast, and not break any part of Cmder (and its dependencies) -- all the qualities of the clink auto updater as I perceive it.

Since @daxgames isn't quite onboard with Cmder having an auto self updater mechanism (https://github.com/cmderdev/cmder/issues/2788#issuecomment-1872628483), I've put implementing this on hold for now; especially since I've not fully addressed all concerns yet, but I'd strongly argue that a portable package like Cmder should indeed have this functionality.  Again, taking clink as inspiration.

Many of the additions I've added to the `.PS1` scripts also have the quality of being used on user's machines to for example fetch the latest versions of each repo and update their "sources.json" script using GitHub's API.  The functionality to download and extract the ZIP file can be added there as well to bring their other parts of the Cmder installation also up to date.

> [...] any local changes the user has made to the scripts will get overwritten. That isn't necessarily a reason against having an auto-updater, but it's something that needs to be communicated clearly to users.

Quite true, and I agree with the "communicated clearly to users" part.

Fortunately, @daxgames have already addressed this about how Cmder's update policy works.  I remember about a decade ago, it was normal for users to download a fresh Cmder package, and then modify the scripts in the `/vendor` directory directly.  This would mean that any modifications they'd do to the code would get overwritten by an update being applied.

@daxgames advocated for clearly separating any "vendor" code from any user "config" code. It is made clear that absolutely **NO** changes must be done by the user to any of the files in the "vendor" directory (unless they make a PR directly in the repo to introduce an upstream change) any modifications they do to that directory isn't guaranteed to be preserved at all.

On the other hand, the "config" directory is the correct place where they are expected to introduce their changes.

There has been a lot of work done by @daxgames to fix this problem, and I'm confident that by now this shouldn't be a problem.  If the user needs extra modifications, we should provide the capability for so, without needing the "vendor" directory to be modified by the user, and instead helping them applying their changes using the "config" directory.

> it's critical that auto-updaters need to validate digital signatures before installing updates. This is not just a theoretical attack vector; other software has been attacked in exactly this way. I changed Clink's auto-updater in v1.9.12 to validate digital signatures before installing an update.

Another great point, this is exactly one of the other reasons that I have put this on hold; originally, I intended to only rely on the validity of the TLS connection made directly to the GitHub API. However, this might not be sufficient, considering the recent attacks, a software package such as Cmder certainly has to be more secure than that if auto updates are implemented.

> If the network connection can be compromised (e.g. man-in-the-middle network attack), or if the file server is hacked (e.g. if GitHub gets hacked)

I would wager that the former is more probable than the latter.  A malicious root CA + DNS spoofing could easily allow an auto update mechanism to download a malicious package and extract it without any additional checks on top of Cmder, so I'm strongly attracted to the idea of implementing digital signatures as a secondary security check for this purpose as well.

I feel like the probability of GitHub's servers getting hacked to serve malicious files is probably lower, at least not in the same capacity that a random shared web hosting (like in the case of Notepad++) might suffer from.  Either way I feel like we would have bigger problems if GitHub servers suffered an attack like this.

In any case, implementing digital signature verification is a 👍🏻 from me.

> if the author's account is hacked

That is indeed a probable security concern; GitHub has attempted to make this harder by requiring 2FA/2SV.

One problem worth considering is that the GitHub actions CI scripts I've implemented regularly fetch updates and commit them directly to the `sources.json` file with the latest release version of these repos; I also personally attempt to do my due diligence in verifying anything that gets added/modified to the Cmder package, but with the amount of code that we already incorporate from Clink and more, it basically just comes to trust in the author to keeping their repos safe.

In any case, I'm very open in improving the security of any mechanisms we already have or plan to add to Cmder.

> Notepad++ put its private key in its public repo

Did they really?! 🤯

> provided that the private key is properly secured.

I believe @MartiUK (the main maintainer of Cmder) might be able to help with that.  I think the Cmder's main domain (`cmder.app`) is also registered by him as well, although I'm not sure.

The bigger concern in my opinion is the cost of the certificate renewal, which could be hundreds of USD, but I'm not exactly sure about that as well.

> A custom updater program (or script) would need to be created

This was originally my intention, writing an updater program in C++ *as part of the launcher) or a PS1 script that could be invoked by the user.

> [...] and validate digital signatures [...] Clink v1.9.12 and newer have an undocumented Lua API that can validate digital signatures 

The idea of re-using the same functionality from Clink's lua code is appealing, instead of writing a second, separate implementation purely for Cmder.

In fact, I was originally going to base this implementation off of the ConEmu's auto update mechanism, since it actively updated at the time.  I wonder if @Maximus5 also considered these security aspects in his implementation.

> It's a medium-sized chunk of work to create a robust auto-updater, and I'm not sure whether I want to sign up for doing that.

It is indeed the case, it takes effort to write a good, secure, performant and feature-rich one.

I planned to study how Clink and ConEmu did it, and replicate that functionality in Cmder as well.  In case it would be possible to share the efforts, I would prefer to contribute to a shared library that then could be re-used across several packages, instead of writing separate pieces of software all intended to do the same thing differently.

The Clink's implementation already seems great, and I'm a fan.  I also was looking forward to re-use ConEmu's code as well, back when I first proposed adding this feature in Cmder.

> The publisher of clink-completions and clink-gizmos (i.e. me) would need to generate a digitally signed Catalog file for each update.
It's a small-sized chunk of work to create a script to automate creating a signed Catalog file for each update. I'm willing to add that to my workflow, but it doesn't accomplish anything without a custom updater program to make use of the signed Catalog file.

Since you're considering the security of the feature, I appreciate and applaud your efforts, so thank you 👍🏻 

I hope we can aim to achieve feature parity in Cmder as well.  It wouldn't make sense for Cmder to have parts of it digitally signed, while the other parts (e.g. the core) not benefitting from the same hardening.

Ideally, those parts could also be digitally signed, and the Cmder project as a whole would benefit from shared efforts in that regard.

**Reply by daxgames** [reply]
Hey guys I just now saw this gigantic thread of discussion it's going to take me awhile to digest it but I am looking at it

**Reply by chrisant996** [reply]
> The bigger concern in my opinion is the cost of the certificate renewal, which could be hundreds of USD, but I'm not exactly sure about that as well.

I pay $35 per year to Certum for a personal cert that adds "Open Source Developer" after my name (discount for OSS devs).  It can be used for as many things as you want (not limited to one project, and no limit on number of signatures).

The first time purchase has an additional charge for a security device to hold the cert.  It does not support putting the cert in a web property (not even as a GitHub secret).  So it's not suitable for a cloud build pipeline, but it's great for local builds. 

https://certum.store/open-source-code-signing.html

**Reply by MartiUK** [reply]
Code signing is insanely expensive prospect for a project that has received one(!) donation in its entire life.

I think it would be better to look at build attestation and verification.

https://docs.github.com/en/actions/how-tos/secure-your-work/use-artifact-attestations/use-artifact-attestations

**Reply by chrisant996** [reply]
> Code signing is insanely expensive prospect for a project that has received one(!) donation in its entire life.

I completely get and respect that $35/year can easily be an impractical cost for an unfunded altruistic project like Cmder.

For my purposes, I don't consider $35/year to be insanely expensive, because I get to use the code signing certs for everything I do.  And also because it lets me gain experience in that space.

Honestly, I think for Cmder a bigger problem than money would be the fact that Certum's code signing certificate for Open Source projects is a single physical security device, and would be super hard to share among a team like in the case of Cmder.

> I think it would be better to look at build attestation and verification.
> 
> https://docs.github.com/en/actions/how-tos/secure-your-work/use-artifact-attestations/use-artifact-attestations

Interesting.  Build attestations seem to be only about provenance, and verifying attestations doesn't verify whether files contain the expected content.  And verification requires the GitHub CLI to be installed on the user's machine.  It appears to be more useful for auditing a pipeline, and not for verifying whether specific content matches an expected state.  (I.e. it can be used to verify what went into a build, but not whether files you have match that build.)

If the downloaded content can't be verified to match the expected content, then I'd recommend against having an auto-updater that _by default_ both downloads files and installs the downloaded files.  It's one thing for malicious code to at least require an explicit user action to get installed, but it's a whole other thing for malicious code be automatically downloaded and installed without any explicit user action (starting a terminal would not be an "explicit user action" with respect to installing updates).
